server/src/main/scala/org/http4s/armeria/server/ArmeriaHttp4sHandler.scala

```diff
@@ -133,8 +133,11 @@ private[armeria] class ArmeriaHttp4sHandler[F[_]](
         if (tail == Stream.empty)
           Pull.done
         else
-          Pull.eval(F.async_[Unit] { cb =>
-            discardReturn(writer.whenConsumed().thenRun(() => cb(RightUnit)))
+          Pull.eval(F.async[Unit] { cb =>
+            F.delay(discardReturn(writer.whenConsumed().thenRun(() => cb(RightUnit))))
+              .as(Some(F.delay(
+                writer.close()
```

Comment:
If `F.delay(writer.close())` is a callback that is called when cancelled, `writer.close()` does not make sense because a response could complete successfully.

If a call is canceled in the middle of processing, `ServiceRequestContext.cancel()` would be a better choice to cancel/reset the HTTP stream.